### PR TITLE
Video dl progress

### PR DESCRIPTION
**Pull request by horahoradev**
This PR:
1. improves scheduler efficiency (only sleeps in the context of no results from postgres, making it 50-100x faster)
2. adds video download progress notifications

## Explanation for #3 
There are many components to this feature. First, I hard-coded a progress hook which publishes to activemq via STOMP on my yt-dlp fork, here: https://github.com/horahoradev/yt-dlp/blob/a9b2c8205eb4ec0781a97c2a859985e1bd5aacb8/yt_dlp/downloader/common.py#L36
It publishes to a video-id specific topic. All subscribers to the topic will receive all mesages.
The credentials are currently hard-coded, which isn't great... but I guess it's fine? Activemq isn't exposed publicly (outside of the docker-compose network).

Next, I added a publish to the state topic within scheduler when the video's state is changed, e.g. here: https://github.com/horahoradev/horahora/blob/video_dl_progress/scheduler/internal/models/video_dl_request.go#L49

The client subscribes to the following topics via websocket:
1. /topic/state
This tells the client when some state has changed, e.g. the video finished downloading (should be removed from the queue), or a new video was added (should be pushed to the end of the list), or a new video started downloading (should be prepended to the list, all downloading videos go on page 1).

2. each per-video topic for progress notifications (e.g. if we downloaded www.nicovideo.jp/watch/sm9 , we'd subscribe to /topic/sm9)
yt-dlp sends progress notifications into the per-video topic, and we receive them directly on the client, setting the progress bar value accordingly.

Permissions are managed via the stomp_proxy service (a new microservice). The client connects directly to nginx, which proxies stomp_proxy, and then stomp_proxy parses each STOMP frame from the client and evaluates whether it's permitted. For now, stomp_proxy simply disallows the client from publishing, and allows all subscribes, which should be fine.

### CODE REVIEW

File: webapp/src/ArchivalPage.js [hunk 1]
```diff
@@ -161,8 +308,20 @@ function ArchivalPage() {
...
             key: 'website',
+        },
+        {
+            title: 'Download Status',
+            key: 'DlStatus',
+            dataIndex: 'DlStatus',
+        },
+        {
+            title: 'Progress',
+            key: 'progress',
```
Comment (sainttttt) [inline webapp/src/ArchivalPage.js:321]: Convention for key naming? Are we starting with uppercase names like with Website and DlStatus?

File: webapp/src/ArchivalPage.js [hunk 2]
```diff
@@ -1,19 +1,164 @@
...
+    
+    }, []);
+
+    function processMessage(message) {
+      mutex.lock();
+      setVideoInProgressDataset(videosInProg => {
+          if (videosInProg == null) {
+              return videosInProg;
+          }
+        let dataset = JSON.parse(JSON.stringify(videosInProg))
```
Comment (sainttttt) [inline webapp/src/ArchivalPage.js:73]: Seems redundant

Comment (horahoradev) [reply webapp/src/ArchivalPage.js:73]: it's a very stupid deep copy mechanism. I might be able to take it out now that I'm using the functional approach to setting this variable.

File: scheduler/main.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func main() {
 
 	m := &sync.Mutex{}
 	// Start three goroutines to subscribe to channel and download items
```
[the remark this answers is not in the document]
Comment (horahoradev) [reply scheduler/main.go:64]: it should just be taken out, yeah, hold on

File: scheduler/internal/models/archive_request_repo.go [hunk 4]
```diff
@@ -181,19 +181,20 @@ func (m *ArchiveRequestRepo) RetryArchivalRequest(userID, downloadID uint64) err
...
-	Website string `db:"website"`
+	ID       string `db:"video_id"`
+	Website  string `db:"website"`
+	DlStatus int    `db:"dlstatus"` // postgres lowercased it, lol
 }
 
 func (m *ArchiveRequestRepo) GetDownloadsInProgress() ([]Video, error) {
 	var videos []Video
-	sql := "select video_id, website FROM videos WHERE dlStatus = 3"
+	sql := "select video_id, website, dlStatus FROM videos WHERE dlStatus >= 3 ORDER BY dlStatus ASC"
```
[the remark this answers is not in the document]
Comment (horahoradev) [reply scheduler/internal/models/archive_request_repo.go:191]: fix't

File: scheduler/internal/schedule/schedule.go [hunk 5]
```diff
@@ -35,10 +39,14 @@ func (p *poller) PollDatabaseAndSendIntoQueue(ctx context.Context, videoQueue ch
 		default:
 			itemsToSchedule, err := p.getVideos()
 			if err != nil {
-				if err != FailedToFetch {
+				if err != sql.ErrNoRows {
 					log.Errorf("failed to get items. Err: %s", err)
+				} else if err == sql.ErrNoRows {
```
[the remark this answers is not in the document]
Comment (horahoradev) [reply scheduler/internal/schedule/schedule.go:44]: hah, yes

File: webapp/src/ArchivalPage.js (hunk 1 above)
Comment (horahoradev) [reply webapp/src/ArchivalPage.js:321]: It turns out there's no need for a key field if we have a dataindex. I switched it to uppercase.

**Pull request merged**